### PR TITLE
Refactor E2E port registration

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Refactoring

**What does this PR do? Why is it needed?**

Currently our port registration in e2e is all over the place. Apart from being hard to maintain, this caused port clashes in #10227. The purpose of this PR is to make the code cleaner:
- explicitly define all base port values in `params.go`
- define a separate `ports` struct
- define a `port` helper function which calculates the actual port numbers to use, taking testing shards into account

**Which issues(s) does this PR fix?**

Unblocks #10277

### CODE REVIEW

File: testing/endtoend/params/params.go [hunk 1]
```diff
@@ -76,27 +96,93 @@ func Init(beaconNodeCount int) error {
 	if !ok {
 		return errors.New("expected TEST_UNDECLARED_OUTPUTS_DIR to be defined")
 	}
-	testIndexStr, ok := os.LookupEnv("TEST_SHARD_INDEX")
+	testTotalShardsStr, ok := os.LookupEnv("TEST_TOTAL_SHARDS")
```
Comment (prestonvanloon) [inline testing/endtoend/params/params.go:99]: Is this a bazel environment variable?

Comment (prestonvanloon) [reply testing/endtoend/params/params.go:99]: nvm it is

Comment (rkapka) [reply testing/endtoend/params/params.go:99]: yes

**Review by rauljordan — COMMENTED** [review]
This is awesome @rkapka. I think the only thing missing is adding a unit test for the `port` function that covers some edge cases

File: testing/endtoend/params/params.go [hunk 2]
```diff
@@ -108,28 +194,125 @@ func InitMultiClient(beaconNodeCount int, lighthouseNodeCount int) error {
...
-		ValidatorGatewayPort:      7150 + testIndex*100,
+		Ports:                     testPorts,
 	}
 	return nil
 }
+
+// port returns a safe port number based on the seed and shard data.
+func port(seed, shardCount, shardIndex int, existingRegistrations *[]int) (int, error) {
+	for _, p := range *existingRegistrations {
+		if seed >= p && seed <= p+(50/shardCount)-1 {
```
Comment (nisdas) [inline testing/endtoend/params/params.go:307]: parametrize the value `50` here

File: testing/endtoend/params/params.go [hunk 3]
```diff
@@ -108,28 +194,125 @@ func InitMultiClient(beaconNodeCount int, lighthouseNodeCount int) error {
...
-		BeaconNodeMetricsPort:     5100 + testIndex*100,
-		ValidatorMetricsPort:      6100 + testIndex*100,
-		ValidatorGatewayPort:      7150 + testIndex*100,
+		Ports:                     testPorts,
 	}
 	return nil
 }
+
+// port returns a safe port number based on the seed and shard data.
+func port(seed, shardCount, shardIndex int, existingRegistrations *[]int) (int, error) {
```
Comment (nisdas) [inline testing/endtoend/params/params.go:305]: a small unit test for this method would be good for general correctness

Comment (rkapka) [reply testing/endtoend/params/params.go:305]: Yeah, it already found a bug...

File: testing/endtoend/params/params.go (hunk 2 above)
Comment (rkapka) [reply testing/endtoend/params/params.go:307]: I will move it to a constant

**Review by nisdas — APPROVED** [review]
